Equality testing for iterables broken

**Which version of Kotest are you using**
4.2.5

Due to the recent changes from https://github.com/kotest/kotest/commit/6bbbfec0cf69089d3d5d9ec0aea848eb055dca90 equality testing for iterables is broken:

```
listOf("hello") shouldNotBe emptyList<String>() // java.lang.AssertionError: [] should not equal ["hello"]
```

this is caused by a missing check in `IterableEq#checkIterableEquality` that the given iterables are the same size. Also, the implementation is not fast because it will create a list in the background (which is then prompty discarded). 

A (untested) alternative implementation could look something like this:

```
    private fun checkIterableEquality(actual: Iterable<*>, expected: Iterable<*>): Throwable? {
        var index = 0
        val actualIterator = actual.iterator()
        val expectedIterator = expected.iterator()
        while (actualIterator.hasNext() && expectedIterator.hasNext()) {
            val t = eq(actualIterator.next(), expectedIterator.next())
            if (t != null) return failure(
                Expected(expected.show()),
                Actual(actual.show()),
                "Elements differ at index $index: "
            )
            index++
        }
        if(actualIterator.hasNext()) {
            return failure(
                Expected(expected.show()),
                Actual(actual.show()),
                "Expected fewer elements"
            )
        }
        if(expectedIterator.hasNext()) {
            return failure(
                Expected(expected.show()),
                Actual(actual.show()),
                "Expected more elements"
            )
        }
        return null
    }
```

--- a/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/Eq.kt
+++ b/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/Eq.kt
@@ -16,21 +16,22 @@ interface Eq<T> {
 /**
  * Locates the applicable [Eq] for the inputs, and invokes it, returning the error if any.
  */
-fun <T: Any?> eq(actual: T, expected: T): Throwable? {
+fun <T : Any?> eq(actual: T, expected: T): Throwable? {
    // if we have null and non null, usually that's a failure, but people can override equals to allow it
    return when {
       actual === expected -> null
       actual == null && expected == null -> null
       actual == null && expected != null && actual != expected -> actualIsNull(expected)
       actual != null && expected == null && actual != expected -> expectedIsNull(actual)
-      actual !=null && expected != null -> when {
+      actual != null && expected != null -> when {
          actual is Map<*, *> && expected is Map<*, *> -> MapEq.equals(actual, expected)
          actual is Throwable && expected is Throwable -> ThrowableEq.equals(actual, expected)
          actual is Regex && expected is Regex -> RegexEq.equals(actual, expected)
          actual is String && expected is String -> StringEq.equals(actual, expected)
          actual is Number && expected is Number -> NumberEq.equals(actual, expected)
-         actual is Iterable<*> && expected is Iterable<*> -> IterableEq.equals(actual, expected)
-         actual is Array<*> && expected is Array<*> -> IterableEq.equals(actual.toList(), expected.toList())
+         IterableEq.isValidIterable(actual) && IterableEq.isValidIterable(expected) -> {
+            IterableEq.equals(IterableEq.asIterable(actual), IterableEq.asIterable(expected))
+         }
          shouldShowDataClassDiff(actual, expected) -> DataClassEq.equals(actual as Any, expected as Any)
          else -> DefaultEq.equals(actual as Any, expected as Any)
       }

--- a/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/IterableEq.kt
+++ b/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/IterableEq.kt
@@ -6,10 +6,36 @@ import io.kotest.assertions.failure
 import io.kotest.assertions.show.show
 
 object IterableEq : Eq<Iterable<*>> {
+
+   /**
+    * Returns true if this [Iterable] is an iterable that is supported by this typeclass.
+    *
+    * There are many third party libs that use Iterable as a base type (Jackson's Json Node, JDK's Path type, to
+    * name but two) and we cannot safely compare these simply using the IterableEq.
+    *
+    * Therefore only explicit types are accepted by our IterableEq.
+    */
+   fun isValidIterable(it: Any): Boolean {
+      return when (it) {
+         is List<*>, is Set<*>, is Array<*>, is Collection<*> -> true
+         else -> false
+      }
+   }
+
+   fun asIterable(it: Any): Iterable<*> {
+      return when (it) {
+         is Array<*> -> it.asList()
+         is List<*> -> it
+         is Set<*> -> it
+         is Collection<*> -> it
+         else -> error("Cannot convert $it to Iterable<*>")
+      }
+   }
+
    override fun equals(actual: Iterable<*>, expected: Iterable<*>): Throwable? {
       return when {
          actual is Set<*> && expected is Set<*> -> checkSetEquality(actual, expected)
-         else -> checkIterableEquality(actual, expected)
+         else -> checkEquality(actual, expected)
       }
    }
 
@@ -36,7 +62,7 @@ object IterableEq : Eq<Iterable<*>> {
       }
    }
 
-   private fun checkIterableEquality(actual: Iterable<*>, expected: Iterable<*>): Throwable? {
+   private fun checkEquality(actual: Iterable<*>, expected: Iterable<*>): Throwable? {
       var index = 0
       val iter1 = actual.iterator()
       val iter2 = expected.iterator()

--- a/kotest-assertions/kotest-assertions-shared/src/jvmTest/kotlin/com/sksamuel/kotest/IterableShouldBeTest.kt
+++ b/kotest-assertions/kotest-assertions-shared/src/jvmTest/kotlin/com/sksamuel/kotest/IterableShouldBeTest.kt
@@ -1,0 +1,70 @@
+package com.sksamuel.kotest
+
+import io.kotest.core.spec.style.DescribeSpec
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.shouldNotBe
+
+class IterableShouldBeTest : DescribeSpec() {
+   init {
+      describe("shouldBe") {
+         it("list / list") {
+            listOf(1, 2, 3) shouldBe listOf(1, 2, 3)
+            listOf(1, 2, 3) shouldNotBe listOf(5, 6, 7)
+            listOf<Int>() shouldBe emptyList()
+            listOf(1) shouldNotBe emptyList<Int>()
+            emptyList<Int>() shouldNotBe listOf(1)
+         }
+         it("list / set") {
+            listOf(1) shouldBe setOf(1)
+            listOf(1, 2) shouldNotBe setOf(2, 1)
+            emptySet<Int>() shouldBe setOf()
+            listOf(1) shouldNotBe emptySet<Int>()
+            setOf(1) shouldNotBe listOf(1, 1, 1)
+            emptySet<Int>() shouldNotBe listOf(1)
+            emptyList<Int>() shouldNotBe setOf(1)
+         }
+         it("set / set") {
+            setOf(1) shouldBe setOf(1)
+            setOf(1) shouldBe setOf(1, 1, 1)
+            setOf(1, 2, 3) shouldBe setOf(3, 2, 1)
+            emptySet<Int>() shouldBe setOf()
+            setOf<Int>() shouldBe emptySet()
+            setOf(1) shouldNotBe emptySet<Int>()
+            emptySet<Int>() shouldNotBe setOf(1)
+            setOf(1, 2) shouldNotBe setOf(1)
+            setOf(1, 2) shouldNotBe setOf(1, 3)
+         }
+         it("linked hash set / list") {
+            LinkedHashSet(setOf(1, 2, 3)) shouldBe listOf(1, 2, 3)
+            listOf(1, 2, 3) shouldBe LinkedHashSet(setOf(1, 2, 3))
+            LinkedHashSet<Int>() shouldBe emptyList()
+            emptyList<Int>() shouldBe LinkedHashSet<Int>()
+         }
+         it("array / array") {
+            arrayOf(1, 2, 3) shouldBe arrayOf(1, 2, 3)
+            arrayOf(1, 2, 3) shouldNotBe arrayOf(5, 6, 7)
+            arrayOf<Int>() shouldBe emptyArray()
+            arrayOf(1) shouldNotBe emptyArray<Int>()
+            emptyArray<Int>() shouldNotBe arrayOf(1)
+         }
+         it("array / list") {
+            arrayOf(1, 2, 3) shouldBe listOf(1, 2, 3)
+            listOf(1, 2, 3) shouldNotBe listOf(5, 6, 7)
+            arrayOf(1, 2, 3) shouldNotBe arrayOf(5, 6, 7)
+            arrayOf<Int>() shouldBe emptyList<Int>()
+            arrayOf(1) shouldNotBe emptyList<Int>()
+            listOf(1) shouldNotBe emptyArray<Int>()
+            emptyArray<Int>() shouldNotBe listOf(1)
+            emptyList<Int>() shouldNotBe arrayOf(1)
+         }
+         it("array / java.util.ArrayList") {
+            arrayOf(1, 2, 3) shouldBe java.util.ArrayList(listOf(1, 2, 3))
+            arrayOf(1, 2, 3) shouldNotBe java.util.ArrayList(listOf(5, 6, 7))
+            arrayOf<Int>() shouldBe java.util.ArrayList<Int>()
+            arrayOf(1) shouldNotBe java.util.ArrayList<Int>()
+            arrayOf(1) shouldNotBe java.util.ArrayList<Int>()
+            emptyArray<Int>() shouldNotBe java.util.ArrayList(listOf(1))
+         }
+      }
+   }
+}

--- a/kotest-assertions/kotest-assertions-shared/src/jvmTest/kotlin/com/sksamuel/kotest/eq/IterableEqTest.kt
+++ b/kotest-assertions/kotest-assertions-shared/src/jvmTest/kotlin/com/sksamuel/kotest/eq/IterableEqTest.kt
@@ -3,9 +3,13 @@ package com.sksamuel.kotest.eq
 import io.kotest.assertions.assertSoftly
 import io.kotest.assertions.eq.IterableEq
 import io.kotest.core.spec.style.FunSpec
+import io.kotest.inspectors.forAll
 import io.kotest.matchers.nulls.shouldBeNull
 import io.kotest.matchers.nulls.shouldNotBeNull
 import io.kotest.matchers.shouldBe
+import io.kotest.matchers.shouldNotBe
+import java.nio.file.Paths
+import java.util.concurrent.ConcurrentLinkedQueue
 import kotlin.time.ExperimentalTime
 import kotlin.time.seconds
 
@@ -100,8 +104,35 @@ class IterableEqTest : FunSpec({
    }
 
    test("should work for empty lists") {
-      IterableEq.equals(emptyList<Int>(), listOf(1))?.message shouldBe "Element 1 expected at index 0 but there were no further elements"
+      IterableEq.equals(emptyList<Int>(), listOf(1))?.message shouldBe
+         "Element 1 expected at index 0 but there were no further elements"
       IterableEq.equals(listOf(1), emptyList<Int>())?.message shouldBe "Unexpected element at index 0: 1"
+   }
+
+   test("shouldNotBe should work for empty lists") {
+      listOf("hello") shouldNotBe emptyList<String>()
+      emptyList<String>() shouldNotBe listOf("hello")
+   }
+
+   test("IterableEq unsupported types") {
+      listOf(
+         Paths.get("foo/bar")
+      ).forAll {
+         IterableEq.isValidIterable(it) shouldBe false
+      }
+   }
+
+   test("IterableEq supported types") {
+      listOf(
+         setOf(1),
+         listOf(1),
+         java.util.ArrayList(listOf(1)),
+         java.util.HashSet(listOf(1)),
+         arrayOf(1),
+         ConcurrentLinkedQueue<Int>(),
+      ).forAll {
+         IterableEq.isValidIterable(it) shouldBe true
+      }
    }
 })
 

